tests/pyfunc/test_spark.py

```diff
@@ -273,3 +396,56 @@ def get_model(_):
     # Running again should see no newly-loaded models.
     results2 = spark.sparkContext.parallelize(range(100), 30).map(get_model).collect()
     assert min(results2) > 0
+
+
+@pytest.mark.skipif(
+    not sys.platform.startswith("linux"),
+    reason="Only Linux system support setting  parent process death signal via prctl lib.",
+)
+@pytest.mark.large
+def test_spark_udf_embedded_model_server_killed_when_job_canceled(spark, sklearn_model, model_path):
+    from mlflow.pyfunc.scoring_server.client import ScoringServerClient
+    from mlflow.models.cli import _get_flavor_backend
+
+    mlflow.sklearn.save_model(sklearn_model.model, model_path)
+
+    server_port = 51234
+
+    @pandas_udf("int")
+    def udf_with_model_server(it: Iterator[pd.Series]) -> Iterator[pd.Series]:
+        from mlflow.models.cli import _get_flavor_backend
+
+        _get_flavor_backend(model_path, no_conda=False, workers=1, install_mlflow=False).serve(
+            model_uri=model_path,
+            port=server_port,
+            host="127.0.0.1",
+            enable_mlserver=False,
+            synchronous=False,
+        )
+
+        time.sleep(120)
+        for x in it:
+            yield x
+
+    def run_job():
+        # Start a spark job with only one UDF task,
+        # and the udf task starts a mlflow model server process.
+        spark.range(1).repartition(1).select(udf_with_model_server("id")).collect()
+
+    _get_flavor_backend(model_path, no_conda=False, install_mlflow=False).prepare_env(
+        model_uri=model_path
+    )
+
+    job_thread = threading.Thread(target=run_job)
+    job_thread.start()
+
+    client = ScoringServerClient("127.0.0.1", server_port)
+    client.wait_server_ready(timeout=20)
+    spark.sparkContext.cancelAllJobs()
+    job_thread.join()
+
+    time.sleep(5)  # waiting server to exit and release the port.
```

Comment:
Is this always enough time to wait to avoid flakiness?

Reply:
increase to 10s. should be long enough. When killed, the server exits quickly.